[caffe2] dont call cudnnDestroy on thread exit (crashes on windows with cuda 11/12)

fuzic opened the pull request:
Summary:
My team has been hitting a mysterious crash for a few months on a windows binary that uses Caffe2 inside a worker thread.

When this thread gets destroyed, there is an error at this line in context_gpu.h where the state of this operation gives CUDNN_STATUS_INTERNAL_ERROR instead of CUDNN_STATUS_SUCCESS.

When enabling cudnn debug logs (via the env variables nvidia specifies), I can see that the context is destroyed twice, even though this code only destroys it once, so something mysterious is causing a double free.

This seems very very similar to the issue/fix described here for pytorch:
https://github.com/pytorch/pytorch/issues/17658
https://github.com/apache/tvm/pull/8267

And pytorch handles this in the same way, by just not calling cudnnDestroy

This seems to have become an issue with cuda11, but I tested cuda12 as well and found that the issue persists so this needs to be somehow fixed.

Test Plan:
CI

I checked that the specific windows binary I am using is able to create and drestroy caffe2-invoking threads without causing the application to crash.

buck run arvr/mode/win/cuda11/opt //arvr/projects/nimble/prod/tools/MonoHandTrackingVis

Differential Revision: D43538017



cc @ngimel

malfet reviewed (approved):
From PyTorch perspective this is a dead code, so sure, looks good to me, though  it's sad that somebody is still using caffe2 (and especially caffe2 on Windows).

Most likely what happens: Windows runtime has DLL_PROCESS_DETACH hook and cuDNN is probably cleans up contexts there. So a better guard would be to just skip this step on Windows (or make cudnnDestory) errors non-throwable

fuzic commented:
@malfet updated with your suggestions - I'm not sure what other approvals I need to get to land this?

fuzic committed:
[caffe2] dont check cudnnDestroy status on thread exit (crashes on windows with cuda 11/12) (#95382)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/95382

My team has been hitting a mysterious crash for a few months on a windows binary that uses Caffe2 inside a worker thread.

When this thread gets destroyed, there is an error at this line in context_gpu.h where the state of this operation gives CUDNN_STATUS_INTERNAL_ERROR instead of CUDNN_STATUS_SUCCESS.

When enabling cudnn debug logs (via the env variables nvidia specifies), I can see that the context is destroyed twice, even though this code only destroys it once, so something mysterious is causing a double free.

This seems very very similar to the issue/fix described here for pytorch:
https://github.com/pytorch/pytorch/issues/17658
https://github.com/apache/tvm/pull/8267

And pytorch handles this in the same way, by just not calling cudnnDestroy

This seems to have become an issue with cuda11, but I tested cuda12 as well and found that the issue persists so this needs to be somehow fixed.

Test Plan:
CI

I checked that the specific windows binary I am using is able to create and drestroy caffe2-invoking threads without causing the application to crash.

buck run arvr/mode/win/cuda11/opt //arvr/projects/nimble/prod/tools/MonoHandTrackingVis

Reviewed By: malfet

Differential Revision: D43538017

fbshipit-source-id: bcec01b8cda7bf7f6ef14324b1ef52379639f511
